protocol/xmrmaker/instance.go

```diff
@@ -57,15 +60,126 @@ func NewInstance(cfg *Config) (*Instance, error) {
 		go cfg.Network.Advertise()
 	}
 
-	return &Instance{
+	inst := &Instance{
 		backend:        cfg.Backend,
 		dataDir:        cfg.DataDir,
 		walletFile:     cfg.WalletFile,
 		walletPassword: cfg.WalletPassword,
 		offerManager:   om,
 		swapStates:     make(map[types.Hash]*swapState),
 		net:            cfg.Network,
-	}, nil
+	}
+
+	err = inst.checkForOngoingSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	return inst, nil
+}
+
+func (b *Instance) checkForOngoingSwaps() error {
+	swaps, err := b.backend.SwapManager().GetOngoingSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swaps {
+		if s.Provides != types.ProvidesXMR {
+			continue
+		}
+
+		if s.Status == types.KeysExchanged || s.Status == types.ExpectingKeys {
+			// TODO: set status to aborted, delete info from recovery db
+			continue
+		}
+
+		err = b.createOngoingSwap(s)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *Instance) createOngoingSwap(s *swap.Info) error {
+	// check if we have shared secret key in db; if so, recover XMR from that
+	// otherwise, create new swap state from recovery info
+	moneroStartHeight, err := b.backend.RecoveryDB().GetMoneroStartHeight(s.ID)
+	if err != nil {
+		return fmt.Errorf("failed to get monero start height for ongoing swap, id %s: %s", s.ID, err)
+	}
+
+	sharedKey, err := b.backend.RecoveryDB().GetSharedSwapPrivateKey(s.ID)
+	if err == nil {
+		b.backend.XMRClient().Lock()
+		defer b.backend.XMRClient().Unlock()
+
+		// TODO: do we want to transfer this back to the original account?
+		addr, err := monero.CreateWallet( //nolint:govet
+			"xmrmaker-swap-wallet",
+			b.backend.Env(),
+			b.backend.XMRClient(),
+			sharedKey,
+			moneroStartHeight,
+		)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("refunded XMR from swap %s: wallet addr is %s", s.ID, addr)
+		return nil
+	}
+
+	offer, err := b.offerManager.GetOfferFromDB(s.ID)
+	if err != nil {
+		return fmt.Errorf("failed to get offer for ongoing swap, id %s: %s", s.ID, err)
+	}
+
+	ethSwapInfo, err := b.backend.RecoveryDB().GetContractSwapInfo(s.ID)
+	if err != nil {
+		return fmt.Errorf("failed to get offer for ongoing swap, id %s: %s", s.ID, err)
+	}
+
+	sk, err := b.backend.RecoveryDB().GetSwapPrivateKey(s.ID)
+	if err != nil {
+		return fmt.Errorf("failed to get private key for ongoing swap, id %s: %s", s.ID, err)
+	}
+
+	relayerInfo, err := b.backend.RecoveryDB().GetSwapRelayerInfo(s.ID)
+	if err != nil {
+		// we can ignore the error; if the key doesn't exist,
+		// then no relayer was set for this swap.
+		relayerInfo = &types.OfferExtra{}
+	}
+
+	b.swapMu.Lock()
+	defer b.swapMu.Unlock()
```

Comment:
Do we need to hold the lock when calling `newSwapStateFromOngoing`? If we only need the lock for one line below, I would just write:
```
b.swapMu.Lock()
b.swapStates[s.ID] = ss
b.swapMu.Unlock()
```
That would be the most self documenting about what requires synchronization if, in fact, it is the only line that needs synchronization.

Reply:
when I wrote this I thought there was a reason we might have needed it during `newSwapStateFromOngoing` but re-reading it, don't think we actually do. updated it!